### PR TITLE
feat: add revocation to access-client and w3up-client

### PR DESCRIPTION
**Pull request by travis**
Add a new method on `access-client`'s `Agent` to make this work and call into it from `w3up-client`'s `Client`.

### CODE REVIEW

File: packages/access-client/test/agent.test.js [hunk 1]
```diff
@@ -249,4 +251,50 @@ describe('Agent', function () {
...
+        revoke: provide(UCAN.revoke, async ({ capability, invocation }) => {
+          // copy a bit of the production revocation handler to do basic validation
+          const { nb: input } = capability
+          const ucan = Delegation.view(
+            { root: input.ucan, blocks: invocation.blocks },
+            // eslint-disable-next-line unicorn/no-null
+            null
+          )
+          return ucan
+            ? { ok: { time: Date.now() } }
```
Comment (alanshaw) [inline packages/access-client/test/agent.test.js:271]: Is this seconds or milliseconds? The interface comment says seconds.

**Review by alanshaw — APPROVED** [review]
Looks ok to me but please resolve the question on the timestamp before merge 🙂

**Review by Gozala — CHANGES_REQUESTED** [review]
looks good but there's on edge case to handle

File: packages/access-client/test/agent.test.js (hunk 1 above)
Comment (travis) [reply packages/access-client/test/agent.test.js:271]: oh I'm sure the interface comment is right! I saw `number`s in the types and didn't check, will fix

File: packages/access-client/src/agent.js [hunk 2]
```diff
@@ -205,6 +206,36 @@ export class Agent {
...
+   * If the delegation was issued by this agent (and therefore is stored in the
+   * delegation store) you can just pass the CID. If not, or if the current agent's
+   * delegation store no longer contains the delegation, you MUST pass a chain of
+   * proofs that proves your authority to revoke this delegation as `options.proofs`.
+   *
+   * @param {import('@ucanto/interface').UCANLink} delegationCID
+   * @param {object} [options]
+   * @param {import('@ucanto/interface').Delegation[]} [options.proofs]
+   */
+  async revokeDelegation(delegationCID, options = {}) {
```
[the remark this answers is not in the document]
Comment (travis) [reply packages/access-client/src/agent.js:221]: happy to defer to your aesthetics here, done!

File: packages/access-client/src/agent.js [hunk 3]
```diff
@@ -205,6 +206,36 @@ export class Agent {
...
+    if (!delegation) {
+      throw new Error(
+        `could not find delegation ${delegationCID.toString()} - please include the delegation in options.proofs`
+      )
+    }
+    return this.invokeAndExecute(UCAN.revoke, {
+      nb: {
+        ucan: delegation.cid,
+      },
+      proofs: options.proofs ?? [delegation],
```
[the remark this answers is not in the document]
Comment (travis) [reply packages/access-client/src/agent.js]: ah good catch thanks! done

File: packages/access-client/test/agent.test.js [hunk 4]
```diff
@@ -249,4 +251,50 @@ describe('Agent', function () {
...
+      audience: fixtures.alice,
+      audienceMeta: {
+        name: 'sss',
+        type: 'app',
+      },
+    })
+
+    const receipt = await agent.revokeDelegation(delegation.cid)
+    assert(receipt.out.ok, `failed to revoke: ${receipt.out.error?.message}`)
+  })
```
[the remark this answers is not in the document]
Comment (travis) [reply packages/access-client/test/agent.test.js:327]: done and done

File: packages/capabilities/src/types.ts [hunk 5]
```diff
@@ -323,6 +332,7 @@ export type UCANRevokeFailure =
   | UCANNotFound
   | InvalidRevocationScope
   | UnauthorizedRevocation
+  | Ucanto.Failure
```
[the remark this answers is not in the document]
Comment (travis) [reply packages/capabilities/src/types.ts]: ah good call, done!
